subsys/net/lib/sockets/sockets_can.c

```diff
@@ -0,0 +1,365 @@
+	 * calling thread (and only the calling thread)
+	 */
+	_k_object_recycle(ctx);
+#endif
+
+	z_finalize_fd(fd, ctx,
+		      (const struct fd_op_vtable *)&can_sock_fd_op_vtable);
+
+	return fd;
+}
+
+static void zcan_received_cb(struct net_context *ctx, struct net_pkt *pkt,
+			     union net_ip_header *ip_hdr,
+			     union net_proto_header *proto_hdr,
+			     int status, void *user_data)
+{
+	NET_DBG("ctx %p pkt %p st %d ud %p", ctx, pkt, status, user_data);
+
+	/* if pkt is NULL, EOF */
+	if (!pkt) {
+		struct net_pkt *last_pkt = k_fifo_peek_tail(&ctx->recv_q);
+
+		if (!last_pkt) {
+			/* If there're no packets in the queue, recv() may
+			 * be blocked waiting on it to become non-empty,
+			 * so cancel that wait.
+			 */
+			sock_set_eof(ctx);
+			k_fifo_cancel_wait(&ctx->recv_q);
+			NET_DBG("Marked socket %p as peer-closed", ctx);
+		} else {
+			net_pkt_set_eof(last_pkt, true);
+			NET_DBG("Set EOF flag on pkt %p", ctx);
+		}
+
+		return;
+	}
+
+	/* Normal packet */
+	net_pkt_set_eof(pkt, false);
+
+	k_fifo_put(&ctx->recv_q, pkt);
+}
+
+static int zcan_bind_ctx(struct net_context *ctx, const struct sockaddr *addr,
+			 socklen_t addrlen)
+{
+	struct sockaddr_can *can_addr = (struct sockaddr_can *)addr;
+	struct net_if *iface;
+	int ret;
+
+	if (addrlen != sizeof(struct sockaddr_can)) {
+		return -EINVAL;
+	}
+
+	iface = net_if_get_by_index(can_addr->can_ifindex);
+	if (!iface) {
+		return -ENOENT;
+	}
+
+	net_context_set_iface(ctx, iface);
+
+	ret = net_context_bind(ctx, addr, addrlen);
+	if (ret < 0) {
+		errno = -ret;
+		return -1;
+	}
+
+	/* For CAN socket, we expect to receive packets after call to bind().
+	 */
+	ret = net_context_recv(ctx, zcan_received_cb, K_NO_WAIT,
+			       ctx->user_data);
+	if (ret < 0) {
+		errno = -ret;
+		return -1;
+	}
+
+	return 0;
+}
+
+ssize_t zcan_sendto_ctx(struct net_context *ctx, const void *buf, size_t len,
+			int flags, const struct sockaddr *dest_addr,
+			socklen_t addrlen)
+{
+	s32_t timeout = K_FOREVER;
+	struct net_pkt *pkt;
+	int ret;
+
+	/* Setting destination address does not probably make sense here so
+	 * ignore it. You need to use bind() to set the CAN interface.
+	 */
+	if (dest_addr) {
+		NET_DBG("CAN destination address ignored");
+	}
+
+	if ((flags & ZSOCK_MSG_DONTWAIT) || sock_is_nonblock(ctx)) {
+		timeout = K_NO_WAIT;
+	}
+
+	pkt = net_pkt_get_tx(ctx, timeout);
+	if (!pkt) {
+		NET_DBG("Cannot get net_pkt (ctx %p)", ctx);
+		errno = EAGAIN;
+		return -1;
+	}
+
+	net_pkt_set_family(pkt, AF_CAN);
+
+	len = net_pkt_append(pkt, len, buf, timeout);
+	if (!len) {
+		net_pkt_unref(pkt);
+		errno = ENOMEM;
+		return -1;
+	}
+
+	/* Register the callback before sending in order to receive the response
+	 * from the bus.
+	 */
+	ret = net_context_recv(ctx, zcan_received_cb, K_NO_WAIT,
+			       ctx->user_data);
+	if (ret < 0) {
+		net_pkt_unref(pkt);
+		errno = -ret;
+		return -1;
+	}
+
+	ret = net_context_sendto(pkt, dest_addr, addrlen, NULL, timeout, NULL,
```

Comment:
you will need to switch to net_context_sendto_new() so you won't build the pkt here in socket, but net_context will

Reply:
I was hoping we could postpone this change to bit later, after all the net_context_sendto() is still there and is working just fine.

Reply:
As you like, but check anyway: I think it is not too much work to switch to net_context_sendto_new , if that's so: please switch.